src/pytest_memray/marks.py

```diff
@@ -267,9 +267,142 @@ def limit_leaks(
     )
 
 
+@dataclass
+class _TrackedObjectsInfo:
+    """Type that holds information about objects that survived tracking."""
+
+    surviving_objects: list
+    num_stacks: int
+    native_stacks: bool
+
+    @property
+    def section(self) -> PytestSection:
+        """Return a tuple in the format expected by section reporters."""
+        body = self._generate_section_text()
+        return (
+            "memray-tracked-objects",
+            "List of leaked objects:\n" + body,
+        )
+
+    @property
+    def long_repr(self) -> str:
+        """Generate a longrepr user-facing error message."""
+        return f"Test leaked {len(self.surviving_objects)} objects"
+
+    def _generate_section_text(self) -> str:
+        """Generate a text summary of leaked objects."""
+        from collections import Counter
+
+        # Group objects by type
+        type_counts = Counter(type(obj).__name__ for obj in self.surviving_objects)
+
+        text_lines = []
+        padding = " " * 4
+
+        # Show top object types that leaked
+        text_lines.append(
+            f"{padding}Object types that leaked (total: {len(self.surviving_objects)} objects):"
+        )
+        for obj_type, count in type_counts.most_common(10):
+            text_lines.append(f"{padding*2}- {obj_type}: {count} instance(s)")
+
+        # Show a sample of the actual objects (up to 10)
+        text_lines.append(f"\n{padding}Sample of leaked objects:")
+        for obj in self.surviving_objects[:10]:
+            try:
+                obj_repr = repr(obj)
+                # Truncate long representations
+                if len(obj_repr) > 100:
+                    obj_repr = obj_repr[:97] + "..."
+                text_lines.append(f"{padding*2}- {type(obj).__name__}: {obj_repr}")
+            except Exception:
+                text_lines.append(f"{padding*2}- {type(obj).__name__}: <repr failed>")
+
+        if len(self.surviving_objects) > 10:
+            text_lines.append(
+                f"{padding*2}... and {len(self.surviving_objects) - 10} more"
+            )
+
+        return "\n".join(text_lines)
+
+
+def track_leaked_objects(
+    _result_file: Path,
+    _config: Config,
+    _test_id: str,
+    _surviving_objects: list | None = None,
+) -> _TrackedObjectsInfo | None:
+    """Track objects that survive the test execution."""
+    if _surviving_objects is None:
+        return None
+
```

Comment:
```suggestion
```
the below `if` already handles everything falsey, so no need to have this check separately.